azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/DeployMojo.java

```diff
@@ -70,26 +77,96 @@
-            final WebApp app = getWebApp();
-            if (app == null) {
-                if (this.isDeployToDeploymentSlot()) {
-                    throw new AzureExecutionException(WEBAPP_NOT_EXIST_FOR_SLOT);
-                }
-                createWebApp(runtimeHandler);
-            } else {
-                updateWebApp(runtimeHandler, app);
-            }
-            deployArtifacts(getWebAppConfiguration());
-        } catch (IOException | AzureAuthFailureException | InterruptedException e) {
-            throw new AzureExecutionException(
-                    String.format("Encountering error when deploying to azure: '%s'", e.getMessage()), e);
+        final WebAppConfig config = getWebAppConfig();
+        final IAppService target = createOrUpdateResource(config);
+        deployArtifacts(target, config);
+    }
+
+    private IAppService createOrUpdateResource(final WebAppConfig config) throws AzureExecutionException {
+        if (StringUtils.isNotEmpty(config.getDeploymentSlotName())) {
+            final IWebApp target = getWebApp(config);
+            return target.exists() ? updateWebApp(target, config) : createWebApp(target, config);
+        } else {
+            final IWebAppDeploymentSlot target = getDeploymentSlot(config);
+            return target.exists() ? updateDeploymentSlot(target, config) : createDeploymentSlot(target, config);
         }
     }
 
+    private IWebApp getWebApp(final WebAppConfig config) {
+        return az.webapp(config.getResourceGroup(), config.getAppName());
+    }
+
+    private IWebAppDeploymentSlot getDeploymentSlot(final WebAppConfig config) throws AzureExecutionException {
+        final IWebApp webApp = getWebApp(config);
+        final boolean isDeploymentSlot = StringUtils.isNotEmpty(config.getDeploymentSlotName());
+        if (isDeploymentSlot && !webApp.exists()) {
+            throw new AzureExecutionException(WEBAPP_NOT_EXIST_FOR_SLOT);
+        }
+        return webApp.deploymentSlot(config.getDeploymentSlotName());
+    }
+
+    private IWebApp createWebApp(final IWebApp webApp, final WebAppConfig webAppConfig) {
+        // todo: get or create Resource Group
+        final ResourceGroup resourceGroup = null;
+        // Get or create App Service Plan
+        final IAppServicePlan appServicePlan = az.appServicePlan(webAppConfig.getServicePlanResourceGroup(), webAppConfig.getServicePlanName());
+        if (!appServicePlan.exists()) {
+            appServicePlan.create()
+                    .withName(webAppConfig.getServicePlanName())
+                    .withResourceGroup(resourceGroup.getName())
+                    .withRegion(webAppConfig.getRegion())
+                    .withPricingTier(webAppConfig.getPricingTier())
+                    .withOperatingSystem(webAppConfig.getRuntime().getOperatingSystem())
+                    .commit();
+        }
+        return webApp.create().withName(webAppConfig.getAppName())
```

Comment:
if appServicePlan exists, shall we verify region, pricingTier, and os?

Reply:
Will add the full logic in later pr